devlib/module/sched.py

```diff
@@ -0,0 +1,262 @@
+# limitations under the License.
+
+import re
+
+from devlib.module import Module
+from devlib.utils.misc import memoized
+
+from past.builtins import basestring
+
+
+class SchedModule(Module):
+
+    name = 'sched'
+
+    sysfs_root = '/sys/devices/system/'
+
+    @staticmethod
+    def probe(target):
+        has_sd = target.directory_exists(SchedDomainData.sched_domain_root)
+        if not has_sd:
+            return False
+
+        return len(target.list_directory(SchedDomainData.sched_domain_root)) > 0
+
+    def get_sd_info(self, cpu):
+        return SchedDomainData(self.target, cpu)
+
+    def get_capacity(self, cpu):
+        """
+        :returns: The capacity of 'cpu'
+        """
+        return self.get_capacities()[cpu]
+
+    @memoized
+    def _has_em(self, cpu):
+        """
+        :returns: Whether energy model data is available for 'cpu'
+        """
+        procfs = SchedDomainData(self.target, cpu).procfs
+        return procfs["domain0"].get("group0", {}).get("energy", {}).get("cap_states") != None
+
+    @memoized
+    def _has_dmips_capacity(self, cpu):
+        """
+        :returns: Whether dmips capacity data is available for 'cpu'
+        """
+        return self.target.file_exists(
+            self.target.path.join(self.sysfs_root, 'cpu/cpu{}/cpu_capacity'.format(cpu))
+        )
+
+    @memoized
+    def _get_em_capacity(self, cpu):
+        """
+        :returns: The maximum capacity value exposed by the EAS energy model
+        """
+        sd = SchedDomainData(self.target, cpu)
+
+        cap_states = sd.domains[0].groups[0].energy.cap_states
+        return int(cap_states.split('\t')[-2])
+
+    @memoized
+    def _get_dmips_capacity(self, cpu):
+        """
+        :returns: The capacity value generated from the capacity-dmips-mhz DT entry
+        """
+        return self.target.read_value(
+            self.target.path.join(
+                self.sysfs_root,
+                'cpu/cpu{}/cpu_capacity'.format(cpu)
+            ),
+            int
+        )
+
+    @memoized
+    def get_capacities(self, default=None):
+        """
+        :param default: Default capacity value to find if no data is
+        found in procfs
+
+        :returns: a dictionnary of the shape {cpu : capacity}
+
+        :raises RuntimeError: Raised when no capacity information is
+        found and 'default' is None
+        """
+        cpus = list(range(self.target.number_of_cpus))
+
+        capacities = {}
+
+        for cpu in cpus:
+            if self._has_em(cpu):
+                capacities[cpu] = self._get_em_capacity(cpu)
+            elif self._has_dmips_capacity(cpu):
+                capacities[cpu] = self._get_dmips_capacity(cpu)
+            else:
+                if default != None:
+                    capacities[cpu] = default
+                else:
+                    raise RuntimeError('No capacity data for cpu{}'.format(cpu))
+
+        return capacities
+
+    @memoized
+    def get_hz(self):
```

Comment:
Don't particularly like the usage of "get_" prefixes... is that to be consistent with other modules?
Otherwise, I would probably like better these to be defined like "properties"...

Reply:
Yeah modules like thermal or cpufreq have loads of set/get methods, so I tried to follow that.

Reply:
It should be either `get_hz` as it is now, or `hz` with an additional `@property` decorator (method names should always be verbs, but properties are data types).